### PR TITLE
UCT/API: return base addr and alloc len from md_mem_query

### PR DESCRIPTION
**Pull request by Akshay-Venkatesh**
## What / Why ?
Allow uct_md_mem_query to return base address and allocation length of the original allocation if the md can detect the memory type for the given buffer. This would allow cuda_cpy md to return base address and allocation length for the UCP registration function to register the whole allocation length as opposed to just user provided length and amortize registration cost for subsequent registration requests for buffers belonging to the same allocation.

### CODE REVIEW

**Comment by Akshay-Venkatesh** [comment]
@yosefe Made UCT API changes based on some of your suggestions offline. Please review when possible.

File: src/ucp/core/ucp_mm.c [hunk 1]
```diff
@@ -111,9 +114,27 @@ ucs_status_t ucp_mem_rereg_mds(ucp_context_h context, ucp_md_map_t reg_md_map,
             } else {
                 ucs_assert(address && length);
 
+                requested_reg_length  = length;
+                requested_reg_address = address;
+
+                if (md_attr->cap.detect_mem_types & UCS_BIT(mem_type)) {
+                    mem_attr.field_mask   = UCT_MD_MEM_ATTR_FIELD_BASE_ADDRESS |
+                                            UCT_MD_MEM_ATTR_FIELD_ALLOC_LENGTH;
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c:122]: but ib md can't detect Cuda memory

File: src/uct/cuda/base/cuda_md.c [hunk 2]
```diff
@@ -145,6 +145,18 @@ UCS_PROFILE_FUNC(ucs_status_t, uct_cuda_base_mem_query,
         mem_attr->mem_type = mem_type;
     }
 
+    if ((mem_attr->field_mask & UCT_MD_MEM_ATTR_FIELD_BASE_ADDRESS) ||
+        (mem_attr->field_mask & UCT_MD_MEM_ATTR_FIELD_ALLOC_LENGTH)) {
```
Comment (yosefe) [inline src/uct/cuda/base/cuda_md.c:149]: according to API, need to check address/length flags individually

File: src/ucp/core/ucp_mm.c [hunk 3]
```diff
@@ -111,9 +114,27 @@ ucs_status_t ucp_mem_rereg_mds(ucp_context_h context, ucp_md_map_t reg_md_map,
...
+
+                    status = uct_md_mem_query(context->tl_mds[md_index].md,
+                                              address, length, &mem_attr);
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c:125]: we need to do this when detecting memory type

File: src/ucp/core/ucp_mm.c (hunk 1 above)
Comment (Akshay-Venkatesh) [reply src/ucp/core/ucp_mm.c:122]: Does it make sense to see if md can register given memory, then search for the md that can detect the given memory and get back attributes and use that to register? Based on this https://github.com/openucx/ucx/pull/6178#discussion_r562901684 if we get this back during memory detection then we have to pass those through all the UCP layers.

File: src/uct/cuda/base/cuda_md.c (hunk 2 above)
Comment (yosefe) [reply src/uct/cuda/base/cuda_md.c:153]: can use single & operation to test both bits

File: src/uct/api/uct.h [hunk 4]
```diff
@@ -1310,15 +1316,25 @@ typedef struct uct_md_mem_attr {
      * Index of the system device on which the buffer resides. eg: NUMA/GPU
      */
     ucs_sys_device_t  sys_dev;
+
+    /**
+     * Base address of the allocation to which the provided buffer belongs to.
+     */
+    void              *base_address;
```
Comment (yosefe) [inline src/uct/api/uct.h:1323]: 1. pls add UCT level gtest for the new API. it has to pass on all transports with some basic implementation.
2. let's split UCP part to another PR.

File: src/ucp/core/ucp_mm.c [hunk 5]
```diff
@@ -111,9 +116,35 @@ ucs_status_t ucp_mem_rereg_mds(ucp_context_h context, ucp_md_map_t reg_md_map,
...
 
+                requested_reg_length  = length;
+                requested_reg_address = address;
+
+                if (md_attr->cap.reg_mem_types & UCS_BIT(mem_type)) {
+
+                    for (detect_md_index = 0; detect_md_index < context->num_mds;
+                         ++detect_md_index) {
+                        detect_md_attr = &context->tl_mds[detect_md_index].attr;
+                        if (detect_md_attr->cap.detect_mem_types & UCS_BIT(mem_type)) {
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c:127]: we need to avoid overhead for memory registration on host memory. consider that most of the times we should go through the registration cache and memtype cache which are relatively fast, so would not want to pay the cost of another pointer detection

Comment (Akshay-Venkatesh) [reply src/ucp/core/ucp_mm.c:127]: Also thought about this concern and makes sense to not pay for this overhead every time registration is requested. As you suggested let's address this part in a separate PR.

File: src/uct/api/uct.h (hunk 4 above)
Comment (Akshay-Venkatesh) [reply src/uct/api/uct.h:1328]: @yosefe removed UCP bits and extended existing gtest. Can you have a look when possible?

File: src/uct/api/uct.h [hunk 6]
```diff
@@ -1309,21 +1309,27 @@ typedef struct uct_md_mem_attr {
...
     /**
      * Index of the system device on which the buffer resides. eg: NUMA/GPU
+     * If the md not support sys_dev query, then UCS_SYS_DEVICE_ID_UNKNOWN
+     * is returned.
      */
     ucs_sys_device_t  sys_dev;
 
     /**
      * Base address of the allocation to which the provided buffer belongs to.
+     * If the md not support base address query, then NULL is returned.
```
Comment (yosefe) [inline src/uct/api/uct.h:1326]: should return same address as passed

File: src/uct/api/uct.h [hunk 7]
```diff
@@ -1309,21 +1309,27 @@ typedef struct uct_md_mem_attr {
...
      */
     void              *base_address;
 
     /**
      * Length of the whole allocation to which the provided buffer belongs to.
+     * If the md does not support querying allocation length then -1 is returned.
```
Comment (yosefe) [inline src/uct/api/uct.h:1332]: should return same length as passed

File: test/gtest/uct/test_md.cc [hunk 8]
```diff
@@ -299,11 +299,15 @@ UCS_TEST_P(test_md, mem_type_detect_mds) {
...
-                              UCT_MD_MEM_ATTR_FIELD_SYS_DEV;
+        mem_attr.field_mask = UCT_MD_MEM_ATTR_FIELD_MEM_TYPE     |
+                              UCT_MD_MEM_ATTR_FIELD_SYS_DEV      |
+                              UCT_MD_MEM_ATTR_FIELD_BASE_ADDRESS |
+                              UCT_MD_MEM_ATTR_FIELD_ALLOC_LENGTH;
         status = uct_md_mem_query(md(), address, buffer_size, &mem_attr);
         ASSERT_UCS_OK(status);
         EXPECT_EQ(alloc_mem_type, mem_attr.mem_type);
+        EXPECT_TRUE((mem_attr.alloc_length >= buffer_size) || (mem_attr.alloc_length == -1));
+        EXPECT_TRUE((mem_attr.base_address == NULL) || (mem_attr.base_address = address));
```
Comment (yosefe) [inline test/gtest/uct/test_md.cc:310]: 1. == vs. =
2. pls really test it : allocate some buffer and check if pointer/size in middle of buffer can find the real address/length of the backing buffer

File: src/uct/cuda/base/cuda_md.c [hunk 9]
```diff
@@ -146,6 +150,25 @@ UCS_PROFILE_FUNC(ucs_status_t, uct_cuda_base_mem_query,
...
+        cu_err = cuMemGetAddressRange(&base_address, &alloc_length,
+                                      (CUdeviceptr)address);
+        if (cu_err != CUDA_SUCCESS) {
+            cuGetErrorString(cu_err, &cu_err_str);
+            ucs_warn("ccuMemGetAddressRange(%p) error: %s", address,
+                     cu_err_str);
+        }
+
+        if (mem_attr->field_mask & UCT_MD_MEM_ATTR_FIELD_BASE_ADDRESS) {
+            mem_attr->base_address = (void *)base_address;
```
Comment (yosefe) [inline src/uct/cuda/base/cuda_md.c:164]: remove space before `*`

File: src/uct/api/uct.h (hunk 6 above)
Comment (Akshay-Venkatesh) [reply src/uct/api/uct.h:1326]: Concerning the above case and this https://github.com/openucx/ucx/pull/6178/#discussion_r564801229. What about the following case?
```
a = malloc(4096); // or cudaMalloc for that matter
uct_md_mem_query(..., address=(a + 32), length=32, ...)
```
It doesn't seem right that mem_attr.base_address return `a + 32` and mem_attr.alloc_length return `32`

File: test/gtest/uct/test_md.cc (hunk 8 above)
Comment (Akshay-Venkatesh) [reply test/gtest/uct/test_md.cc:310]: 1. Will fix!
2. Unfortunately I couldn't test as on the current machine, I have all gtests fail because max transports exceeded issue. And gtests don't accept UCX_TLS and I believe UCX_TLS is always hardcoded to `all`. The other machines where I have smaller number of NICs should be available soon. I'll try with change `1.` and test and push changes.

Comment (Akshay-Venkatesh) [reply test/gtest/uct/test_md.cc]: Pushed changes to query at random offests and things work with CUDA. 
```
[----------] 1 test from cuda_cpy/test_md
[ RUN      ] cuda_cpy/test_md.mem_type_detect_mds/0 <cuda_cpy>
[     INFO ] cuda: sys_dev[] (0000:06:00.0)
[     INFO ] cuda-managed: sys_dev[] (0000:06:00.0)
[       OK ] cuda_cpy/test_md.mem_type_detect_mds/0 (147 ms)
[----------] 1 test from cuda_cpy/test_md (147 ms total)
```
ROCM md is untested though.

Comment (yosefe) [reply test/gtest/uct/test_md.cc]: 2. can try with UCX_NET_DEVICES

File: src/uct/api/uct.h (hunk 6 above)
Comment (Akshay-Venkatesh) [reply src/uct/api/uct.h:1326]: @yosefe how should we proceed here?

File: test/gtest/uct/test_md.cc [hunk 10]
```diff
@@ -303,11 +305,20 @@ UCS_TEST_P(test_md, mem_type_detect_mds) {
...
-        ASSERT_UCS_OK(status);
-        EXPECT_EQ(alloc_mem_type, mem_attr.mem_type);
-        EXPECT_TRUE((mem_attr.alloc_length >= buffer_size) || (mem_attr.alloc_length == -1));
-        EXPECT_TRUE((mem_attr.base_address == NULL) || (mem_attr.base_address = address));
+
+        for (unsigned i = 0; i < 300; i++) {
+            offset        = ucs::rand() % buffer_size;
+            length_offset = ucs::rand() % buffer_size;
+
+            status = uct_md_mem_query(md(), (void*)((char*)address + offset),
```
Comment (yosefe) [inline test/gtest/uct/test_md.cc:313]: UCS_PTR_BYTE_OFFSET

File: test/gtest/uct/test_md.cc [hunk 11]
```diff
@@ -303,11 +305,20 @@ UCS_TEST_P(test_md, mem_type_detect_mds) {
...
+                                      length_offset, &mem_attr);
+            ASSERT_UCS_OK(status);
+            EXPECT_EQ(alloc_mem_type, mem_attr.mem_type);
+            EXPECT_TRUE((mem_attr.alloc_length >= buffer_size) ||
```
Comment (yosefe) [inline test/gtest/uct/test_md.cc:317]: why not == ?

File: src/uct/api/uct.h (hunk 6 above)
Comment (yosefe) [reply src/uct/api/uct.h:1326]: why doesn't it make sense? it will simplify the code which tries to "expand" the memory registration

Comment (Akshay-Venkatesh) [reply src/uct/api/uct.h:1326]: If registration is the only user of uct_md_mem_query with BASE_ADDR, ALLOC_LENGTH attributes requested, then the suggestion to return the pointer and length as is is not erroneous. But if there are other users, then isn't it misleading? Even for registration, ideally we want to return `a` as based address and `4096` (or greater in case the actual allocation results in a larger allocation than user requested) otherwise we're not expanding when there's an option to do so.

File: test/gtest/uct/test_md.cc (hunk 11 above)
Comment (Akshay-Venkatesh) [reply test/gtest/uct/test_md.cc:317]: I was concerned about the case where the underlying allocator may decide to allocate larger range (say because it only allocates at page granularity). I can change this to `==` as it would work for CUDA.

File: src/uct/api/uct.h (hunk 6 above)
Comment (yosefe) [reply src/uct/api/uct.h:1326]: i think the notion of "base address" is not well defined enough to require that base address/length are page multiples.
it should be defined more loosely, as a "continuous memory region which contains address/length and has same memory properties". IMO, anything more strict than that will be hard to define in a transport-independent way

File: test/gtest/uct/test_md.cc (hunk 11 above)
Comment (yosefe) [reply test/gtest/uct/test_md.cc:317]: we can adjust our expectation according to the memory type

File: src/uct/api/uct.h (hunk 6 above)
Comment (Akshay-Venkatesh) [reply src/uct/api/uct.h:1326]: That makes sense. I'll make changes as per your suggestion.

Comment (yosefe) [reply src/uct/api/uct.h:1326]: thanks, let's also document this in the API.

File: test/gtest/uct/test_md.cc (hunk 11 above)
Comment (Akshay-Venkatesh) [reply test/gtest/uct/test_md.cc]: I've changed to use `==` in the PR.

File: src/uct/api/uct.h (hunk 6 above)
Comment (Akshay-Venkatesh) [reply src/uct/api/uct.h]: Addressed this.

File: src/uct/api/uct.h (hunk 7 above)
Comment (Akshay-Venkatesh) [reply src/uct/api/uct.h]: Addressed this.

File: test/gtest/uct/test_md.cc [hunk 12]
```diff
@@ -310,13 +310,15 @@ UCS_TEST_P(test_md, mem_type_detect_mds) {
...
                                       length_offset, &mem_attr);
             ASSERT_UCS_OK(status);
             EXPECT_EQ(alloc_mem_type, mem_attr.mem_type);
-            EXPECT_TRUE((mem_attr.alloc_length >= buffer_size) ||
-                        (mem_attr.alloc_length == -1));
-            EXPECT_TRUE((mem_attr.base_address == NULL) ||
+            EXPECT_TRUE((mem_attr.alloc_length == buffer_size) ||
+                        (mem_attr.alloc_length == length_offset));
+            EXPECT_TRUE((mem_attr.base_address ==
+                         UCS_PTR_BYTE_OFFSET(address, offset)) ||
```
Comment (yosefe) [inline test/gtest/uct/test_md.cc:321]: if mem type is cuda, can we expect `mem_attr.base_address == UCS_PTR_BYTE_OFFSET(address, offset)` ?

File: src/uct/rocm/base/rocm_base.c [hunk 13]
```diff
@@ -220,6 +220,14 @@ ucs_status_t uct_rocm_base_mem_query(uct_md_h md, const void *addr,
         mem_attr_p->sys_dev = UCS_SYS_DEVICE_ID_UNKNOWN;
     }
 
+    if (mem_attr_p->field_mask & UCT_MD_MEM_ATTR_FIELD_BASE_ADDRESS) {
+        mem_attr_p->base_address = NULL;
+    }
+
+    if (mem_attr_p->field_mask & UCT_MD_MEM_ATTR_FIELD_ALLOC_LENGTH) {
+        mem_attr_p->alloc_length = -1;
+    }
```
Comment (yosefe) [inline src/uct/rocm/base/rocm_base.c:229]: need to update these according to the revised API definition

File: test/gtest/uct/test_md.cc [hunk 14]
```diff
@@ -299,11 +301,26 @@ UCS_TEST_P(test_md, mem_type_detect_mds) {
...
-        ASSERT_UCS_OK(status);
-        EXPECT_EQ(alloc_mem_type, mem_attr.mem_type);
+        mem_attr.field_mask = UCT_MD_MEM_ATTR_FIELD_MEM_TYPE     |
+                              UCT_MD_MEM_ATTR_FIELD_SYS_DEV      |
+                              UCT_MD_MEM_ATTR_FIELD_BASE_ADDRESS |
+                              UCT_MD_MEM_ATTR_FIELD_ALLOC_LENGTH;
+
+        for (unsigned i = 0; i < 300; i++) {
+            offset        = ucs::rand() % buffer_size;
+            length_offset = ucs::rand() % buffer_size;
```
Comment (yosefe) [inline test/gtest/uct/test_md.cc:311]: will it work if length_offset==0?
also rename
offset -> slice_offset
length -> slice_length

File: test/gtest/uct/test_md.cc (hunk 12 above)
Comment (Akshay-Venkatesh) [reply test/gtest/uct/test_md.cc:321]: @yosefe I don't think so ([Refer](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__MEM.html#group__CUDA__MEM_1g64fee5711274a2a0573a789c94d8299b)). The second part of this condition check works for cuda. You would see base_address set to the start of the region allocated by cuda.

File: test/gtest/uct/test_md.cc (hunk 14 above)
Comment (Akshay-Venkatesh) [reply test/gtest/uct/test_md.cc:311]: I see that we require uct_md_mem_query to be provided with non-zero lengths as per UCT API documentation. Shall I skip if length_offset==0?

Comment (yosefe) [reply test/gtest/uct/test_md.cc:311]: yes, or add 1 to rand result

File: test/gtest/uct/test_md.cc (hunk 12 above)
Comment (yosefe) [reply test/gtest/uct/test_md.cc:321]: right, so in case of cuda mem_attr.base_address==address? can we check this explicitly?

Comment (Akshay-Venkatesh) [reply test/gtest/uct/test_md.cc:321]: Sure. I can add that check.

Comment (Akshay-Venkatesh) [reply test/gtest/uct/test_md.cc]: @yosefe I reverted the above change because to handle cases where `cuMemGetAddressRange` fails. In this case, I've changed logic in uct cuda_copy to also return provided address and length. I hope that's ok.

File: src/uct/cuda/base/cuda_md.c [hunk 15]
```diff
@@ -85,11 +85,12 @@ UCS_PROFILE_FUNC(ucs_status_t, uct_cuda_base_mem_query,
     uint32_t is_managed        = 0;
     unsigned value             = 1;
     CUdevice cuda_device       = -1;
+    void *base_address         = (void*)address;
```
Comment (yosefe) [inline src/uct/cuda/base/cuda_md.c:88]: no need to cast

File: test/gtest/uct/test_md.cc [hunk 16]
```diff
@@ -307,23 +307,33 @@ UCS_TEST_P(test_md, mem_type_detect_mds) {
                               UCT_MD_MEM_ATTR_FIELD_ALLOC_LENGTH;
 
         for (unsigned i = 0; i < 300; i++) {
-            offset        = ucs::rand() % buffer_size;
-            length_offset = ucs::rand() % buffer_size;
+            slice_offset = ucs::rand() % buffer_size;
+            slice_length = ucs::rand() % buffer_size;
+
+            if (!slice_length) {
```
Comment (yosefe) [inline test/gtest/uct/test_md.cc:313]: == 0

File: test/gtest/uct/test_md.cc [hunk 17]
```diff
@@ -307,23 +307,33 @@ UCS_TEST_P(test_md, mem_type_detect_mds) {
...
+                                                          slice_offset),
+                                      slice_length, &mem_attr);
             ASSERT_UCS_OK(status);
             EXPECT_EQ(alloc_mem_type, mem_attr.mem_type);
             EXPECT_TRUE((mem_attr.alloc_length == buffer_size) ||
-                        (mem_attr.alloc_length == length_offset));
+                        (mem_attr.alloc_length == slice_length));
             EXPECT_TRUE((mem_attr.base_address ==
-                         UCS_PTR_BYTE_OFFSET(address, offset)) ||
+                         UCS_PTR_BYTE_OFFSET(address, slice_offset)) ||
```
Comment (yosefe) [inline test/gtest/uct/test_md.cc:326]: if mem-type is cuda:
EXPECT_EQ(address, mem_attr.base_address)
else:
EXPECT_EQ(UCS_PTR_BYTE_OFFSET(address, slice_offset), mem_attr.base_address)

File: src/uct/cuda/base/cuda_md.c (hunk 15 above)
Comment (Akshay-Venkatesh) [reply src/uct/cuda/base/cuda_md.c:88]: Needed this cast from `const void*` to `void*`.

File: test/gtest/uct/test_md.cc [hunk 18]
```diff
@@ -320,11 +320,15 @@ UCS_TEST_P(test_md, mem_type_detect_mds) {
...
             EXPECT_EQ(alloc_mem_type, mem_attr.mem_type);
-            EXPECT_TRUE((mem_attr.alloc_length == buffer_size) ||
-                        (mem_attr.alloc_length == slice_length));
-            EXPECT_TRUE((mem_attr.base_address ==
-                         UCS_PTR_BYTE_OFFSET(address, slice_offset)) ||
-                        (mem_attr.base_address == address));
+            if ((alloc_mem_type == UCS_MEMORY_TYPE_CUDA) ||
+                (alloc_mem_type == UCS_MEMORY_TYPE_CUDA_MANAGED)) {
+                EXPECT_TRUE(mem_attr.alloc_length == buffer_size);
+                EXPECT_TRUE(mem_attr.base_address == address);
```
Comment (yosefe) [inline test/gtest/uct/test_md.cc:326]: EXPECT_EQ

File: test/gtest/uct/test_md.cc [hunk 19]
```diff
@@ -320,11 +320,15 @@ UCS_TEST_P(test_md, mem_type_detect_mds) {
...
+            } else {
+                EXPECT_TRUE(mem_attr.alloc_length == slice_length);
+                EXPECT_TRUE(mem_attr.base_address ==
```
Comment (yosefe) [inline test/gtest/uct/test_md.cc:329]: EXPECT_EQ

**Comment by Akshay-Venkatesh** [comment]
@yosefe errors seem unrelated to the PR.

File: test/gtest/uct/test_md.cc [hunk 20]
```diff
@@ -322,12 +322,12 @@ UCS_TEST_P(test_md, mem_type_detect_mds) {
             EXPECT_EQ(alloc_mem_type, mem_attr.mem_type);
             if ((alloc_mem_type == UCS_MEMORY_TYPE_CUDA) ||
                 (alloc_mem_type == UCS_MEMORY_TYPE_CUDA_MANAGED)) {
-                EXPECT_TRUE(mem_attr.alloc_length == buffer_size);
-                EXPECT_TRUE(mem_attr.base_address == address);
+                EXPECT_EQ(mem_attr.alloc_length, buffer_size);
+                EXPECT_EQ(mem_attr.base_address, address);
```
Comment (yosefe) [inline test/gtest/uct/test_md.cc:326]: need to inverse order of args

File: test/gtest/uct/test_md.cc [hunk 21]
```diff
@@ -322,12 +322,12 @@ UCS_TEST_P(test_md, mem_type_detect_mds) {
...
             } else {
-                EXPECT_TRUE(mem_attr.alloc_length == slice_length);
-                EXPECT_TRUE(mem_attr.base_address ==
-                            UCS_PTR_BYTE_OFFSET(address, slice_offset));
+                EXPECT_EQ(mem_attr.alloc_length, slice_length);
+                EXPECT_EQ(mem_attr.base_address,
+                          UCS_PTR_BYTE_OFFSET(address, slice_offset));
```
Comment (yosefe) [inline test/gtest/uct/test_md.cc:330]: need to inverse order of args

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@Akshay-Venkatesh pls squash

**Comment by yosefe** [comment]
@shamisp @tonycurtis can you pls take a look?

**Review by tonycurtis — CHANGES_REQUESTED** [review]
uct.h:1283-1288,1325,1332: drop the last "to" in these comments
1311,1318: add period to sentence to separate from next line
1319,1326,1333: If the md does not support ...

File: src/uct/api/uct.h [hunk 22]
```diff
@@ -1303,22 +1309,40 @@ typedef struct uct_md_mem_attr {
```
Comment (bureddy) [inline src/uct/api/uct.h:1309]:
```
"Note that the field mask is
     * populated upon return from uct_md_mem_query and not set by user."
```
is this text correct?

Comment (Akshay-Venkatesh) [reply src/uct/api/uct.h:1309]: Thanks for the catch @bureddy It's not true. I'll address this.

**Comment by shamisp** [comment]
👍

**Review by yosefe — APPROVED** [review]
pls squash

**Review by yosefe — APPROVED** [review]
